### DISCUSSION TITLE
Enum `value` property when a custom metaclass is used

### DISCUSSION THREAD

**Discussion by Viicos** [microsoft/pyright#7939] [Q&A] [answered]
The following commit: https://github.com/microsoft/pyright/commit/7888d1c4ef35a14803389671043c904fc9cb51a9 refering to this issue: https://github.com/microsoft/pyright/issues/5092 made it so that the following would evaluate the value to `Any`:

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.12&code=GYJw9gtgBApgdgV2gSwgBzCALlAooiAGjwIFkYsBDYgSTi3yQChRIosBPNZOAcylQZsUAIJwOxADLIsMEJQA2xSgGcVcrAH1OaGEyYBjBapVQAwgAswyAzBXkqACkYQHlAJQAuJlF9QAdIH6hsZqUHSyvHKW1rYqAOoyFm7OBMQQFJRGJgC8MTZ2bl4%2BfgACaOC62BwlvgAmMMBQAG6KCDCO6grA7lAAtAB8AvSeAUEhJuH0MFEg%2BXGJWFYIWCkuxX5Q5ZUaNZsNTa0K7Z0w3b2Dw1ijgf7B2WGkHC6LyZmOETPRVgUJSUXeTYiKA5KAARn0D1MTxeSTAKxSn1m8zsr3hq0yGz8wNBEP0Jg02i4HRhBFebn8In8R3axDEHHcTAJ2CJukcpKQaIRmUp1LaMCkMjkigA2mCALqM-QHKDABBwRwAdySmgyVFGHIg5MyxGVS3RqsyGueZLh3KoFyGADkwHAYIC-MytDoOnqLIaqHzjgLROJGZsnazXWatGrKF7aVBpLJ5AoxZKmEA)

```python
from enum import Enum, EnumMeta, IntEnum
from typing import Any, Literal, assert_type

class ChoicesMeta(EnumMeta):
    ...


class IntegerChoicesWithMeta(Enum, metaclass=ChoicesMeta):
    @property
    def value(self) -> int: ...

class IntegerChoicesWithoutMeta(Enum):
    @property
    def value(self) -> int: ...


class MyEnumWithMeta(IntegerChoicesWithMeta):
    A = 1

class MyEnumWithoutMeta(IntegerChoicesWithoutMeta):
    A = 1


assert_type(MyEnumWithMeta.A.value, Any)
assert_type(MyEnumWithoutMeta.A.value, Literal[1])


def fun(with_meta: MyEnumWithMeta, without_meta: MyEnumWithoutMeta) -> None:
    assert_type(with_meta.value, Any)
    assert_type(without_meta.value, Literal[1])
```

Which could make sense as said in the linked issue, the metaclass (in the case of the Django `ChoicesMeta` metaclass).

Would it be possible to infer `value` as an `int` if the property is explicitly set on one of the enum classes?

For some context I'm trying to improve the stubs of the Django enums here: https://github.com/typeddjango/django-stubs/pull/2162. I had to go through the internals of Python (and the extra magic Django is doing) enums -- I believe you had to do the same for the draft chapter in the spec -- and was surprised by how much metaclasses (and meta programming) are pushed to their limits. So I would understand if this request is not accepted

**Comment by erictraut** [comment, marked as answer]
Yes, this sounds reasonable. It's also consistent with mypy's current behavior.

The next version of pyright will include this change.
